compiler/rustc_expand/src/expand.rs

```diff
@@ -432,6 +432,14 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         let (mut fragment_with_placeholders, mut invocations) =
             self.collect_invocations(input_fragment, &[]);
 
+        if matches!(fragment_with_placeholders, AstFragment::Crate(_))
+            && self.cx.sess.opts.unstable_opts.input_stats
+        {
+            let mut counter = rustc_ast_passes::node_count::NodeCounter::new();
+            fragment_with_placeholders.visit_with(&mut counter);
+            eprintln!("Pre-expansion node count:  {}", counter.count);
```

Comment:
Pre-expansion genuinely is when only the crate root has been parsed. `mod foo;` is handled kind of like a macro which expands to `mod foo { ... }`. It is handled by the macro expander as macros themself can expand to a `mod foo;`.

Reply:
@bjorn3 I understand that, but what you're saying seems to be not _quite_ accurate, unless I'm misunderstanding something: inside the macroexpander, it calls `collect_invocations` _first_, which actually does the loading of external modules, and then once all the macro invocations are gathered, they're all expanded. That's why the "pre-expansion node count" is actually significantly different before vs after this PR, while still being different from the "post-expansion node count" in both cases.

Still I'm not yet super familiar with how macroexpansion works; I'm guessing it's also possible that a macro can expand to produce a `mod foo;`, in which case `fully_expand_fragment` would get called again and actually result in more external module loading after the first time?

In any case, stepping back a bit, the reason I went down this rabbit hole and opened the PR in the first place is that I genuinely did want to know how many AST nodes there were after all the source code had been parsed but before macroexpansion. And so unless people are actively depending on `-Z input-stats` to give just stats about the crate root, I still just think it's more useful to change the behavior. I started a Zulip thread about this if you'd rather discuss more there: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/Should.20.60-Z.20input-stats.60.20and.20.60-Z.20hir-stats.60.20both.20exist.3F/near/482308786

Reply:
I think that would make sense. Maybe they should be implemented as counters that are incremented during parsing or right after a source file has been parsed and then report the value of the counters _after_ expansion.

Reply:
Sounds good; I'll close this PR and open that followup when I have the chance.